cartographer/mapping_3d/pose_graph/landmark_cost_function.h

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CARTOGRAPHER_MAPPING_3D_POSE_GRAPH_LANDMARK_COST_FUNCTION_H_
+#define CARTOGRAPHER_MAPPING_3D_POSE_GRAPH_LANDMARK_COST_FUNCTION_H_
+
+#include "Eigen/Core"
+#include "Eigen/Geometry"
+#include "cartographer/mapping/pose_graph.h"
+#include "cartographer/mapping_3d/pose_graph/spa_cost_function.h"
+#include "cartographer/transform/rigid_transform.h"
+#include "cartographer/transform/transform.h"
+#include "ceres/ceres.h"
+#include "ceres/jet.h"
+
+namespace cartographer {
+namespace mapping_3d {
+namespace pose_graph {
+
+// Cost function measuring weighted error between the observed relative pose
+// given by sensor and relative pose computed from the linearly interpolated
+// pose of the robot at the moment of observation.
+class LandmarkCostFunction {
+ public:
+  using LandmarkObservation =
+      mapping::PoseGraph::LandmarkNode::LandmarkObservation;
+
+  static ceres::CostFunction* CreateAutoDiffCostFunction(
+      const LandmarkObservation& observation, common::Time prev_node_time,
+      common::Time next_node_time) {
+    return new ceres::AutoDiffCostFunction<
+        LandmarkCostFunction, 6 /* residuals */,
+        3 /* previous node translation variables */,
+        4 /* previous node rotation variables */,
+        3 /* next node translation variables */,
+        4 /* next node rotation variables */,
+        3 /* landmark translation variables */,
+        4 /* landmark rotation variables */>(
+        new LandmarkCostFunction(observation, prev_node_time, next_node_time));
+  }
+
+  template <typename T>
+  bool operator()(const T* const prev_node_rotation,
```

Comment:
This order does not match the template arguments.

Reply:
Thanks!